File: AI_Engine_4_0/src/AI_Engine/AIEngine.java
package AI_Engine;

import java.awt.event.MouseEvent;
import java.awt.event.MouseListener;
import java.awt.event.MouseMotionListener;
import java.awt.event.MouseWheelEvent;
import java.awt.event.MouseWheelListener;

import javax.swing.JButton;
import javax.swing.JFileChooser;
import javax.swing.JFrame;

public class AIEngine {

	static int Layers = 1;
	static int Inputs = 0;
	static int Outputs = 0;
	static int Hidden = 0;
	static int AI_NR = 1;
	static int CON_NR = 0;
	static int GENSTEP_NR = 1;
	static int generationStep = 0;
	static int generation = 0;
	static int DNA_Per = 50;

	int m_x = 0, m_y = 0, l_m_x = 0, l_m_y = 0;
	JFrame jf1;
	JButton save;
	JButton saveE;
	JButton load;
	ANNDisplayer lbl1;
	String[] tempFile;
	boolean overrideWait;

	static AI[] AIs;
	Evolver evolver;
	public DataManager DataManager;

	public int getGeneration() {
		return generation;
	}

	public int getGenerationStep() {
		return generationStep;
	}

	public int getAINumberPerScoreList(int Number) {
		return Evolver.score_list[Number];
	}

	public float getAIScorePerScoreList(int Number) {
		return Evolver.tempPoints[Evolver.score_list[Number]];
	}

	public void openANNDisplayer() {
		GUI();
	}

	public void saveAI(int AI) {
		DataManager.saveAI(AI);
	}

	void extractData(String[] File) {
		int Inputs = Integer.valueOf(File[0]);
		int Outputs = Integer.valueOf(File[1]);
		int Hidden = Integer.valueOf(File[2]);
		int Layers = Integer.valueOf(File[3]);
		int CON_NR = Integer.valueOf(File[4]);
		int DNA_Per = Integer.valueOf(File[5]);
		int GENSTEP_NR = Integer.valueOf(File[6]);
		int AI_NR = Integer.valueOf(File[7]);
		int generation = Integer.valueOf(File[8]);
		int generationStep = Integer.valueOf(File[9]);
		float[][] CON = new float[AI_NR][CON_NR];
		int temp = 0;
		for (int i = 0; i < AI_NR; i++) {
			for (int j = 0; j < CON_NR; j++) {
				CON[i][j] = Float.valueOf(File[temp + 10]);
				temp++;
			}
		}
		overrideEngine(Inputs, Outputs, Hidden, Layers, AI_NR, GENSTEP_NR, DNA_Per, generation, generationStep, CON);
	}

	void overrideEngine(int Inputnumber, int Outputnumber, int Hiddennumber, int Layernumber, int AInumber,
			int Generationsteps, int MutationPerCent, int Generation, int GenerationStep, float[][] CON) {
		Inputs = Inputnumber;
		Outputs = Outputnumber;
		Hidden = Hiddennumber;
		Layers = Layernumber;
		AI_NR = AInumber;
		GENSTEP_NR = Generationsteps;
		DNA_Per = MutationPerCent;
		generation = Generation;
		generationStep = GenerationStep;

		DataManager = new DataManager();
		evolver = new Evolver();
		AIs = new AI[AI_NR];
		if (CON == null) {
			CON_NR = Inputs * Hidden + Hidden * Outputs;
			if (Layers > 1) {
				for (int i = 0; i < Layers - 1; i++) {
					CON_NR += Hidden * Hidden;
				}
			}
			for (int i = 0; i < AI_NR; i++) {
				AIs[i] = new AI(CON_NR, GENSTEP_NR, Inputs, Outputs, i);
			}
		}else {
			CON_NR = Inputs * Hidden + Hidden * Outputs;
			if (Layers > 1) {
				for (int i = 0; i < Layers - 1; i++) {
					CON_NR += Hidden * Hidden;
				}
			}
			for (int i = 0; i < AI_NR; i++) {
				AIs[i] = new AI(CON_NR, GENSTEP_NR, Inputs, Outputs, i);
					AIs[i].setCON(CON[i]);
			}
		}
		
	}

	public void setupEngine(int Inputnumber, int Outputnumber, int Hiddennumber, int Layernumber, int AInumber,
			int Generationsteps, int MutationPerCent) {
		overrideEngine(Inputnumber, Outputnumber, Hiddennumber, Layernumber, AInumber, Generationsteps, MutationPerCent,
				0, 0, null);
		GUI();

	}

	public void setInput(int Input, float Value, int AI) {
		AIs[AI].
		setInput(Input, Value);
	}

	public void runEngine(int AI) {
		AIs[AI].run();
	}

	public float getOutput(int Output, int AI) {
		return AIs[AI].getOutput(Output);
	}

	public void setScore(float Score, int AI) {
		AIs[AI].setScore(Score);
	}

	public void nextGenerationStep() {
		if (generationStep < GENSTEP_NR - 1) {
			generationStep++;
		} else {
			if (overrideWait) {
				extractData(tempFile);
				tempFile = null;
				overrideWait = false;
			} else {
				generationStep = 0;
				evolver.scoreLister();
				evolver.evolve();
				generation++;
			}
		}
	}

	void GUI() {

		jf1 = new JFrame();
		jf1.setSize(800, 800);
		jf1.setLocationRelativeTo(null);
		jf1.setLayout(null);
		jf1.setTitle("Best ANN");
		jf1.addMouseMotionListener(new MouseMotionListener() {

			@Override
			public void mouseDragged(MouseEvent arg0) {
				m_x = (int) arg0.getX();
				m_y = (int) arg0.getY();
				ANNDisplayer.p_x += (m_x - l_m_x) * (1 / ANNDisplayer.zoom);
				ANNDisplayer.p_y += (m_y - l_m_y) * (1 / ANNDisplayer.zoom);
				l_m_x = m_x;
				l_m_y = m_y;

			}

			@Override
			public void mouseMoved(MouseEvent arg0) {
				l_m_x = m_x;
				l_m_y = m_y;
				m_x = (int) arg0.getX();
				m_y = (int) arg0.getY();

			}

		});
		jf1.addMouseWheelListener(new MouseWheelListener() {

			@Override
			public void mouseWheelMoved(MouseWheelEvent arg0) {
				float zoomLevel = arg0.getWheelRotation() * 0.1f;
				ANNDisplayer.zoom -= zoomLevel / (1 / ANNDisplayer.zoom);
				if (ANNDisplayer.zoom > 1) {
					ANNDisplayer.zoom = 1;
				}
			}

		});
		jf1.setResizable(true);
		jf1.requestFocus();
		jf1.setVisible(true);

		save = new JButton("save Best AI");
		save.setVisible(true);
		save.setBounds(13, 13, 120, 25);
		save.addMouseListener(new MouseListener() {

			@Override
			public void mouseReleased(MouseEvent e) {
				// TODO Auto-generated method stub

			}

			@Override
			public void mousePressed(MouseEvent e) {
				// TODO Auto-generated method stub

			}

			@Override
			public void mouseExited(MouseEvent e) {
				// TODO Auto-generated method stub

			}

			@Override
			public void mouseEntered(MouseEvent e) {

				// TODO Auto-generated method stub

			}

			@Override
			public void mouseClicked(MouseEvent e) {
				DataManager.saveAI(0);
				System.out.println("Saved");

			}
		});

		saveE = new JButton("save current Enginestate");
		saveE.setVisible(true);
		saveE.setBounds(143, 13, 180, 25);
		saveE.addMouseListener(new MouseListener() {

			@Override
			public void mouseReleased(MouseEvent e) {
				// TODO Auto-generated method stub

			}

			@Override
			public void mousePressed(MouseEvent e) {

			}

			@Override
			public void mouseExited(MouseEvent e) {
				// TODO Auto-generated method stub

			}

			@Override
			public void mouseEntered(MouseEvent e) {
				// TODO Auto-generated method stub

			}

			@Override
			public void mouseClicked(MouseEvent e) {
				DataManager.saveCurrentEngineState();

			}
		});

		load = new JButton("getEnginestate");
		load.setVisible(true);
		load.setBounds(332, 13, 125, 25);
		load.addMouseListener(new MouseListener() {

			@Override
			public void mouseReleased(MouseEvent e) {
				// TODO Auto-generated method stub

			}

			@Override
			public void mousePressed(MouseEvent e) {
				// TODO Auto-generated method stub

			}

			@Override
			public void mouseExited(MouseEvent e) {
				// TODO Auto-generated method stub

			}

			@Override
			public void mouseEntered(MouseEvent e) {
				// TODO Auto-generated method stub

			}

			@Override
			public void mouseClicked(MouseEvent e) {
				final JFileChooser fc = new JFileChooser();
				/*
				 * FileFilter filter = new FileNameExtensionFilter(".cos", "cos");
				 * fc.setFileFilter(filter);
				 */
				int response = fc.showOpenDialog(load);
				if (response == JFileChooser.APPROVE_OPTION) {
					tempFile = DataManager.getEngineState(fc.getSelectedFile().toString());
					lbl1 = null;
					load = null;
					saveE = null;
					save = null;
					jf1 = null;
					overrideWait = true;
					
				}

			}
		});

		lbl1 = new ANNDisplayer();
		lbl1.setVisible(true);
		lbl1.setBounds(0, 0, 8000, 8000);

		jf1.add(load);
		jf1.add(saveE);
		jf1.add(save);
		jf1.add(lbl1);

	}

}

File: AI_Engine Beta/ClassLibrary1/AIEngine.cs
using System;
using System.Collections.Generic;
using System.IO;
using MathNet.Numerics.LinearAlgebra;
using MathNet.Numerics.LinearAlgebra.Double;

namespace AI_Engine
{
    public class AIEngine
    {
        public static AI[] AIs;

        public static int Inputs;
        public static int Outputs;
        public static int[] hidden;
        public static int AI_NR;
        static int GENSTEP_NR;
        public static int MR;

        public static int generationStep;
        static int generation;
        static Evolver evolver;
        static DataManager DataManager = new DataManager();

        public int getGeneration()
        {
            return generation;
        }

        public int getGenerationStep()
        {
            return generationStep;
        }

        public int getAINumberPerScoreList(int Number)
        {
            return evolver.score_list[Number];
        }

        public float getAIScorePerScoreList(int Number)
        {
            return evolver.tempPoints[evolver.score_list[Number]];
        }

        public void saveAI(int AI)
        {
            DataManager.saveAI(AI);
        }

        public void setupEngine(Boolean menu, int InputNumber, int[] hiddenDimension, int OutputNumber, int AINumber, int Generationsteps,
                int MutationRate)
        {
            if (!menu)
            {

                Inputs = InputNumber;
                hidden = hiddenDimension;
                Outputs = OutputNumber;
                AI_NR = AINumber;
                GENSTEP_NR = Generationsteps;
                MR = MutationRate;
                evolver = new Evolver();
                AIs = new AI[AI_NR];
                for (int i = 0; i < AI_NR; i++)
                {
                    AIs[i] = new AI(Generationsteps, InputNumber, hidden, OutputNumber);
                }
            }
        }

        public void setInput(int Input, double Value, int AI)
        {
            AIs[AI].setInput(Input, Value);
        }

        public void runEngine(int AI)
        {
            AIs[AI].run();
        }

        public double getOutput(int Output, int AI)
        {
            return AIs[AI].getOutput(Output);
        }

        public void setScore(float Score, int AI)
        {
            AIs[AI].setScore(Score);
        }

        public void nextGenerationStep()
        {
            if (generationStep < GENSTEP_NR - 1)
            {
                generationStep++;
            }
            else
            {
                generationStep = 0;
                evolver.scoreLister();
                evolver.evolve();
                generation++;
            }
        }
    }

    public class AI
    {
        private Matrix<double> Inputs;
        private Matrix<double>[] baises;
        private Matrix<double>[] weights;
        private Matrix<double> Outputs;
        private float[] points;

        private double sigmoid(double x)
        {
            return 2 / (1 + Math.Exp(-2 * x)) - 1;
        }

        public AI(int Generationsteps, int Inputs, int[] hidden, int Outputs)
        {
            points = new float[Generationsteps];
            Random random = new Random();
            this.Inputs = DenseMatrix.Create(1, Inputs, random.NextDouble() * 2 - 1);
            baises = new Matrix<double>[hidden.Length + 1];
            weights = new Matrix<double>[hidden.Length + 1];
            int last_layer = Inputs;
            for (int i = 0; i < hidden.Length; i++)
            {
                baises[i] = DenseMatrix.Create(1, hidden[i], random.NextDouble() * 2 - 1);
                weights[i] = DenseMatrix.Create(last_layer, hidden[i], random.NextDouble() * 2 - 1);
                last_layer = hidden[i];

            }
            baises[hidden.Length] = DenseMatrix.Create(1, Outputs, random.NextDouble() * 2 - 1);
            weights[hidden.Length] = DenseMatrix.Create(last_layer, Outputs, random.NextDouble() * 2 - 1);
        }

        public AI(int Generationsteps, int Inputs, Matrix<double>[] baises, Matrix<double>[] weights)
        {
            points = new float[Generationsteps];
            this.Inputs = Matrix<double>.Build.Random(1, Inputs);
            this.baises = baises;
            this.weights = weights;
        }

        public void setInput(int Input, double Value)
        {
            Inputs.At(0, Input, Value);
        }

        public void run()
        {
            Matrix<double> last_layer = Inputs;
            for (int i = 0; i < weights.Length; i++)
            {
                last_layer = last_layer.Multiply(weights[i]);
                last_layer = last_layer.Add(baises[i]);
                for (int j = 0; j < last_layer.ColumnCount; j++)
                {
                    last_layer.At(0, j, sigmoid(last_layer[0, j]));
                }
            }
            Outputs = last_layer;
        }

        public double getOutput(int Output)
        {
            return Outputs[0, Output];
        }

        public void setScore(float Score)
        {
            points[AIEngine.generationStep] = Score;
        }

        public float getAvarageScore()
        {
            float temp = 0;
            for (int i = 0; i < points.Length; i++)
            {
                temp += points[i];
            }
            return temp / points.Length;
        }

        public Matrix<double>[] getWeights()
        {
            return weights;
        }

        public Matrix<double>[] getBaises()
        {
            return baises;
        }

        public void setWeights(Matrix<double>[] weights)
        {
            this.weights = weights;
        }

        public void setBaises(Matrix<double>[] baises)
        {
            this.baises = baises;
        }
    }

    public class Evolver
    {
        public float[] tempPoints;
        public int[] score_list;

        public Evolver()
        {
            tempPoints = new float[AIEngine.AI_NR];
            score_list = new int[AIEngine.AI_NR];
        }

        public void scoreLister()
        {
            float score = 0;
            float score_2 = 0;
            int score_array = 0;
            Boolean[] score_check = new Boolean[AIEngine.AI_NR];

            for (int i = 0; i < AIEngine.AI_NR; i++)
            {
                tempPoints[i] = AIEngine.AIs[i].getAvarageScore();
                score_check[i] = false;
            }
            score = 99999999999999999999999999999999999999f;

            for (int i = 0; i < AIEngine.AI_NR; i++)
            {
                for (int j = 0; j < AIEngine.AI_NR; j++)
                {
                    if (score_2 <= tempPoints[j] && tempPoints[j] <= score && score_check[j] == false)
                    {
                        score_2 = tempPoints[j];
                        score_array = j;
                    }
                }
                score_check[score_array] = true;
                score = tempPoints[score_array];
                score_2 = 0;
                score_list[i] = score_array;
            }
        }

        public void evolve()
        {
            Random random = new Random();
            int parent = 0;
            for (int i = 0; i < AIEngine.AI_NR / 2; i++)
            {
                Matrix<double>[] baises = new Matrix<double>[AIEngine.hidden.Length + 1];
                Matrix<double>[] weights = new Matrix<double>[AIEngine.hidden.Length + 1];
                int last_layer = AIEngine.Inputs;
                for (int j = 0; j < AIEngine.hidden.Length; j++)
                {
                    baises[j] = DenseMatrix.Create(1, AIEngine.hidden[j], 0);
                    weights[j] = DenseMatrix.Create(last_layer, AIEngine.hidden[j], 0);
                    for (int k = 0; k < AIEngine.hidden[j]; k++)
                    {
                        if (random.Next(100) <= 50)
                        {
                            baises[j].At(0, k, AIEngine.AIs[score_list[parent]].getBaises()[j][0, k]);
                        }
                        else
                        {
                            baises[j].At(0, k, AIEngine.AIs[score_list[parent + 1]].getBaises()[j][0, k]);
                        }
                        if (random.Next(100) <= AIEngine.MR)
                        {
                            baises[j].At(0, k, random.Next() * 2 - 1);
                        }
                        for (int l = 0; l < last_layer; l++)
                        {
                            if (random.Next(100) <= 50)
                            {
                                weights[j].At(l, k, AIEngine.AIs[score_list[parent]].getWeights()[j][l, k]);
                            }
                            else
                            {
                                weights[j].At(l, k, AIEngine.AIs[score_list[parent + 1]].getWeights()[j][l, k]);
                            }
                            if (random.Next(100) <= AIEngine.MR)
                            {
                                weights[j].At(l, k, random.NextDouble() * 2 - 1);
                            }
                        }
                    }
                    last_layer = AIEngine.hidden[j];
                }
                baises[AIEngine.hidden.Length] = DenseMatrix.Create(1, AIEngine.Outputs, 0);
                weights[AIEngine.hidden.Length] = DenseMatrix.Create(last_layer, AIEngine.Outputs, 0);
                for (int j = 0; j < AIEngine.Outputs; j++)
                {
                    if (random.Next(100) <= 50)
                    {
                        baises[AIEngine.hidden.Length].At(0, j, AIEngine.AIs[score_list[parent]].getBaises()[AIEngine.hidden.Length][0, j]);
                    }
                    else
                    {
                        baises[AIEngine.hidden.Length].At(0, j, AIEngine.AIs[score_list[parent + 1]].getBaises()[AIEngine.hidden.Length][0, j]);
                    }
                    if (random.Next(100) <= AIEngine.MR)
                    {
                        baises[AIEngine.hidden.Length].At(0, j, random.NextDouble() * 2 - 1);
                    }
                    for (int k = 0; k < last_layer; k++)
                    {
                        if (random.Next(100) <= 50)
                        {
                            weights[AIEngine.hidden.Length].At(k, j, AIEngine.AIs[score_list[parent]].getWeights()[AIEngine.hidden.Length][k, j]);
                        }
                        else
                        {
                            weights[AIEngine.hidden.Length].At(k, j, AIEngine.AIs[score_list[parent + 1]].getWeights()[AIEngine.hidden.Length][k, j]);
                        }
                        if (random.Next(100) <= AIEngine.MR)
                        {
                            weights[AIEngine.hidden.Length].At(k, j, random.NextDouble() * 2 - 1);
                        }
                    }
                }


                AIEngine.AIs[score_list[i + (AIEngine.AI_NR / 2)]].setBaises(baises);
                AIEngine.AIs[score_list[i + (AIEngine.AI_NR / 2)]].setWeights(weights);
                if (parent > AIEngine.AI_NR / 2)
                {
                    parent = 0;
                }
                else
                {
                    parent += 2;
                }
            }
        }
    }

    public class DataManager
    {
        public void saveAI(int AI)
        {
            List<String> content = new List<String>();
            String name = AI + ". AI (" + AIEngine.Inputs + ", " + AIEngine.Outputs + ", " + AIEngine.hidden + ").txt";

            content.Add(AIEngine.Inputs.ToString());
            content.Add(AIEngine.Outputs.ToString());
            content.Add(AIEngine.hidden.Length.ToString());
            for (int i = 0; i < AIEngine.hidden.Length; i++)
            {
                content.Add(AIEngine.hidden[i].ToString());
            }
            Matrix<double>[] weights = AIEngine.AIs[AI].getWeights();
            for (int i = 0; i < weights.Length; i++)
            {
                for (int j = 0; j < weights[i].ColumnCount; j++)
                {
                    for (int k = 0; k < weights[i].RowCount; k++)
                    {
                        content.Add(weights[i][k, j].ToString());
                    }
                }
            }
            Matrix<double>[] baises = AIEngine.AIs[AI].getBaises();
            for (int i = 0; i < baises.Length; i++)
            {
                for (int j = 0; j < baises[i].ColumnCount; j++)
                {

                    content.Add(baises[i][0, j].ToString());

                }
            }
            File.WriteAllText(name, content.ToString());
        }
    }
}
